src/main/java/teammates/ui/controller/Action.java

```diff
@@ -225,6 +225,9 @@ private boolean isSessionTokenValid(String actualToken) {
         if (sessionId == null) {
             // Newly-created session
             sessionId = session.getId();
+        } else {
+            // The requested session ID is appended with .worker-name, which defaults to node0
+            sessionId = sessionId.replaceFirst("\\.node0$", "");
```

Comment:
Any documentation for this information?
Is this fix for dev server only or for all environments?

Reply:
https://www.eclipse.org/jetty/documentation/9.4.x/session-configuration-housekeeper.html
I've tested staging and it seems to work fine, but it's not a fool-proof guarantee.

Reply:
> This is a rather implementation dependent solution.

Which is why I said the fix is not fool-proof.

P.S. we are already implicitly using double-submit cookie technique, just that we do not make use of the token sent via cookie. This is double-submit cookie:
```patch
private boolean isSessionTokenValid(String actualToken) {
-   String sessionId = request.getRequestedSessionId();
-   if (sessionId == null) {
-       // Newly-created session
-       sessionId = session.getId();
-   }
-
-   String expectedToken = CryptoHelper.computeSessionToken(sessionId);
+   String expectedToken = HttpRequestHelper.getCookieValueFromRequest(request, Const.ParamsNames.SESSION_TOKEN);

    return actualToken.equals(expectedToken);
}
```
Maybe it's a false sense of insecurity for me, but relying on values purely generated on the front-end does not seem to be as safe as using something generated on the back-end.

Reply:
`getRequestedSessionId` is also controlled by front-end (client).  http://www.servlets.com/javadoc/javax/servlet/http/HttpServletRequest.html#getRequestedSessionId()

Maybe we should go with the changes you suggested so we can get rid of the implementation dependent removal of `\\.node0$`

Double submit cookie pattern is not safe under two scenarios according to this:
https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Double_Submit_Cookie

And none of the two scenarios apply to our case (usage of subdomain and HTTP instead of HTTPS).